### ISSUE TITLE
Lots of unimplemented arithmetic methods

### ISSUE THREAD

**Issue by NiklasGustafsson** [dotnet/TorchSharp#923]
Marked by 'Obsolete("not implemented", true)' throughout the code, throwing 'NotImplementedException'

**Comment by NiklasGustafsson** [comment]
@MovGP0 -- thanks for adding these and thus highlighting them. I'll get on it! :-)

**Comment by MovGP0** [comment]
Great!

btw: Wanted to implement `torch.matrix_rank`, but it was moved to `torch.linalg.matrix_rank`. 
Not sure if I should sipmly remove `torch.matrix_rank` or just make it a wrapper for `torch.linalg.matrix_rank`.

**Comment by NiklasGustafsson** [comment]
The TorchSharp API "design" is one of pragmatism -- the overall goal is to make it easy to move existing PyTorch code to C#, so duplication like that is fine. If there is Python code that uses `torch.matrix_rank` the by all means keep it in. I added `torch.softmax()` which is not in the docs, but does the same thing as `torch.special.softmax()'. Since its used in Python code, it's better to have it in.

**Comment by GeorgeS2019** [comment]
> The TorchSharp API "design" is one of pragmatism -- the overall goal is to make it easy to move existing PyTorch code to C#

TorchSharp API's Pragmatism is key to growing .NET deep learning community and keeping .NET users within .NET when dealing with deep learning. 

This goal is in POSITIVE progress!

**Comment by NiklasGustafsson** [comment]
Here's an interesting situation:

The doc on `resolve_neg` refers to `is_neg`, which is not document. Does anyone know what it does?

https://pytorch.org/docs/stable/generated/torch.resolve_neg.html#torch.resolve_neg

**Comment by MovGP0** [comment]
> `is_neg`, which is not documented

I found this code in `TensorImpl.h`:

```c
  /**
   * Whether or not the tensor should be negated
   */
  inline bool is_neg() const {
    constexpr auto negative_ks = DispatchKeySet(DispatchKey::Negative);
    return key_set_.has_all(negative_ks);
  }
```

**Comment by NiklasGustafsson** [comment]
BTW, @MovGP0 -- 0.99.3 is out, with most of the reported methods implemented.

**Comment by MovGP0** [comment]
@NiklasGustafsson thanks for the great work you are doing here! 💐
